File: pinyougou_parent/pinyougou_cart_web/src/main/java/com/pinyougou/cart/controller/CartController.java
package com.pinyougou.cart.controller;

import com.alibaba.dubbo.config.annotation.Reference;
import com.pinyougou.cart.service.CartService;
import entity.Result;
import groupEntity.Cart;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import util.CookieUtil;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import java.util.List;

@RestController
@RequestMapping("/cart")
public class CartController {

    @Reference
    private CartService cartService;

    @Autowired
    private HttpSession session;

    @Autowired
    private HttpServletRequest request;

    @Autowired
    private HttpServletResponse response;


    private String getSessionId() {
        String sessionId = CookieUtil.getCookieValue(request, "cartCookie", "utf-8");
        if (sessionId == null) {
            sessionId = session.getId();
            CookieUtil.setCookie(request, response, "cartCookie", sessionId, 3600 * 24 * 7, "utf-8");
        }
        return sessionId;
    }

    @RequestMapping("/findCartList")
    public List<Cart> findCartList() {
        String userName= SecurityContextHolder.getContext().getAuthentication().getName();
        System.out.println(userName);
        String sessionId = getSessionId();
        List<Cart> cartList_sessionId = cartService.selectCartListFromRedis(sessionId);
        if (userName.equals("anonymousUser")){
            return cartList_sessionId;
        }else{
            List<Cart> cartList_userName = cartService.selectCartListFromRedis(userName);
            if (cartList_sessionId!=null&&cartList_sessionId.size()>0){
                cartList_userName = cartService.mergeCartList(cartList_sessionId,cartList_userName);
                cartService.deleteCartList(sessionId);
                cartService.saveCartListToRedis(userName,cartList_userName);
            }
            return cartList_userName;
        }
    }

    @RequestMapping("/addItemToCartList")
    public Result addItemToCartList(Long itemId, Integer num) {
        try {
            String userName= SecurityContextHolder.getContext().getAuthentication().getName();
            String sessionId = getSessionId();
            List<Cart> cartList = findCartList();
            cartList = cartService.addItemToCartList(cartList, itemId, num);
            if (userName.equals("anonymousUser")){
                System.out.println("save cartList to redis by sessionId...");
                cartService.saveCartListToRedis(sessionId,cartList);
            }else{
                System.out.println("save cartList to redis by username...");
                cartService.saveCartListToRedis(userName,cartList);
            }
            return new Result(true,"添加购物车成功");
        } catch (RuntimeException e) {
            e.printStackTrace();
            return new Result(false,e.getMessage());
        }catch (Exception e) {
            e.printStackTrace();
            return new Result(false,"添加购物车失败");
        }
    }
}

File: pinyougou_parent/pinyougou_cart_web/src/main/webapp/js/controller/cartController.js
//控制层
app.controller('cartController', function ($scope, $controller, cartService) {

    $controller('baseController', {$scope: $scope});//继承

    //读取列表数据绑定到表单中
    $scope.findCartList = function () {
        cartService.findCartList().success(
            function (response) {
                $scope.cartList = response;
                sum();
            }
        );
    }
    $scope.addItemToCartList = function (itemId, num) {
        cartService.addItemToCartList(itemId, num).success(function (response) {
            if (response.success) {
                $scope.findCartList();
            } else {
                alert(response.message);
            }
        })
    }

    sum = function () {
        $scope.totalNum = 0;
        $scope.totalMoney = 0.00;
        for (var i = 0; i < $scope.cartList.length; i++) {
            var cart = $scope.cartList[i];
            var orderItemList = cart.orderItemList;
            for (var j = 0; j < orderItemList.length; j++) {
                $scope.totalNum+=orderItemList[j].num;
                $scope.totalMoney+=orderItemList[j].totalFee;
            }
        }
    }
});
